src/backend/renderer/damage/shaper.rs

```diff
@@ -0,0 +1,407 @@
+                    acc.0.min(rect.loc.x),
+                    acc.1.min(rect.loc.y),
+                    acc.2.max(rect.loc.x + rect.size.w),
+                    acc.3.max(rect.loc.y + rect.size.h),
+                    acc.4.max(area),
+                )
+            });
+
+        let bbox_w = x_max - x_min;
+        let bbox_h = y_max - y_min;
+
+        let damage_bbox = Rectangle::<i32, Physical>::from_loc_and_size((x_min, y_min), (bbox_w, bbox_h));
+
+        // Damage the current bounding box when there's a damage rect covering near all the area.
+        if max_damage_area as f32 / (damage_bbox.size.w * damage_bbox.size.h) as f32
+            > MAX_DAMAGE_TO_DAMAGE_BBOX_RATIO
+        {
+            self.out_damage.push(damage_bbox);
+            return;
+        }
+
+        // Now we try to split bounding box to process non-overlapping damage rects separately.
+        //
+        // The whole approach is recursive and splits viewport if and only if we have a gap
+        // in the current segment or rectangles touch each other, since the renderer excludes
+        // borders.
+        //
+        // Examples:
+        //      [0, 3], [1, 2] [2, 3] [3, 4]
+        //      will have a split at ^
+        //      [0, 3], [1, 2] [2, 3] [6, 10]
+        //      will have a split at ^
+        //
+        //  Resulting in recursively trying to shape damage before and after
+        //  split point.
+
+        let mut direction = if bbox_w >= bbox_h {
+            DamageSplitAxis::X
+        } else {
+            DamageSplitAxis::Y
+        };
+
+        if invert_direction {
+            direction = direction.invert();
+        }
+
+        // The coordinate where the first rectangle ends and where the potential overlap
+        // will end.
+        let mut overlap_end = match direction {
+            DamageSplitAxis::X => {
+                if Some(direction) != last_direction {
+                    in_damage.sort_unstable_by(|lhs, rhs| {
+                        // Sort ascending by X and then descending by width, so the first
+                        // rectangle will overlap the most.
+                        lhs.loc.x.cmp(&rhs.loc.x).then(rhs.size.w.cmp(&lhs.size.w))
+                    });
+                }
+                in_damage[0].loc.x + in_damage[0].size.w
+            }
+            DamageSplitAxis::Y => {
+                if Some(direction) != last_direction {
+                    in_damage.sort_unstable_by(|lhs, rhs| {
+                        // Sort ascending by Y and then descending by height, so the first
+                        // rectangle will overlap the most.
+                        lhs.loc.y.cmp(&rhs.loc.y).then(rhs.size.h.cmp(&lhs.size.h))
+                    });
+                }
+                in_damage[0].loc.y + in_damage[0].size.h
+            }
+        };
+
+        // The start of overlap.
+        let mut overlap_start_idx = 0;
+        for idx in overlap_start_idx + 1..in_damage.len() {
```

Comment:
```suggestion
        for idx in 1..in_damage.len() {
```
It's 0, might as well.

Reply:
yeah, but it's more clear that way.